### PR TITLE
Add deferUpdate to GearChangeFrame.

### PR DESCRIPTION
**Pull request by phulin**
This should make a substantial performance improvement. We will now update the equipment lists only once when doing an equipment update, which happens every adventure.

I've also changed `ensureValidSelections` to `updateAllModels` as I believe that's a more consistent and descriptive name.

### CODE REVIEW

**Comment by phulin** [comment]
Ah... seems like I'm not supposed to be initializing a frame on a test. Not sure how to get around that with the way the code is currently structured.

**Comment by jaadams5** [comment]
>This should make a substantial performance improvement.

Any chance you could get before and after profile data that are comparable enough to support this assertion?

**Comment by phulin** [comment]
Yep, will do.

**Comment by phulin** [comment]
Ran the following test a few times and saw between a 20 and 40% decrease in execution time with the deferred updates turned on. Still need to figure out the test issue.

```java
      for (var entry : ItemDatabase.entrySet()) {
        KoLConstants.inventory.add(ItemPool.get(entry.getKey(), 1));
      }
      var outfitOne =
          List.of(...)
              .stream()
              .map(id -> id < 0 ? EquipmentRequest.UNEQUIP : ItemPool.get(id, 1))
              .toList();
      var outfitTwo =
          List.of(...)
              .stream()
              .map(id -> id < 0 ? EquipmentRequest.UNEQUIP : ItemPool.get(id, 1))
              .toList();

      EquipmentManager.setEquipment(outfitOne.toArray(new AdventureResult[outfitOne.size()]));
      EquipmentManager.setEquipment(outfitTwo.toArray(new AdventureResult[outfitTwo.size()]));
      var start = Instant.now();
      for (var i = 0; i < 10; i++) {
        EquipmentRequest.switchEquipment(
            outfitTwo.toArray(new AdventureResult[outfitTwo.size()]),
            outfitOne.toArray(new AdventureResult[outfitOne.size()]));
        EquipmentRequest.switchEquipment(
            outfitOne.toArray(new AdventureResult[outfitOne.size()]),
            outfitTwo.toArray(new AdventureResult[outfitTwo.size()]));
      }
      var end = Instant.now();
      System.out.println("Elapsed: " + Duration.between(start, end).toMillis());
```

**Comment by midgleyc** [comment]
Looks like it might be https://arbitrary-but-fixed.net/2022/01/21/headless-gui-github-actions.html?

**Comment by Veracity0** [comment]
You cannot make a JFrame unless you have a real graphics environment.
However, you can make JPanels and test them. We do that in UseItemEnqueuePanelTest, for example.

What if GearChangeFrame was essentially just a container for GearChangePanel.
I did that for SynthesizeFrame.
```
public class SynthesizeFrame extends GenericFrame {
  public SynthesizeFrame() {
    super("Sweet Synthesis");

    JPanel centerPanel = new JPanel(new BorderLayout());

    centerPanel.add(new SynthesizePanel(), BorderLayout.CENTER);
    centerPanel.add(new StatusPanel(), BorderLayout.SOUTH);

    this.setCenterComponent(centerPanel);
  }
}
```
If all the "guts" of GearChangeFrame is in GearChangePanel, you could instantiate the latter and test it.

There would be other refactoring needed. GearChangeFrame.updateEquipmentModels calls a couple of SwingUtilities methods. It wants to call updateEquipmentModelsInternal() in the SwingEventThread. But, seems to me, it could invoke it directly if GraphicsEnvironment.isHeadless().

File: src/net/sourceforge/kolmafia/swingui/panel/GearChangePanel.java [hunk 1]
```diff
@@ -0,0 +1,1497 @@
+package net.sourceforge.kolmafia.swingui.panel;
+
+import java.awt.*;
```
Comment (midgleyc) [inline src/net/sourceforge/kolmafia/swingui/panel/GearChangePanel.java:3]: we expand the *s on src/ files (or we try to, at least)

**Review by midgleyc — DISMISSED** [review]
Also, the tabs have gone from the top to the bottom. Was there a reason for this change?

**Comment by Veracity0** [comment]
Nice. checked out the PR so I could easily look at what you did with GearChangePanel.

1) GenericFrame had support for tabs. So, obviously you needed to handle them yourself in your panel.
- GenericFrame had support for tabbing forward and back.
I am fine with not having that for this 2-tab panel.
- GenericFrame had support for remembering which tab you were on when you closed it.
Again, I am fine with not doing that; let it always start with the Equipment tab, not the Customizable tab.
- Please put the tabs at SwingConstants.TOP
2) I like the deferredUpdateLevel concept. Reminds me of ConcoctionDatabase refreshes. ;)

And this is outside the scope of this PR, but I wish we had implemented some things as Listeners.
For example, when you add a familiar - or free the king or break ronin in a path that restricts familiars - I wish we could just signal "(familiars)" and both the GearChangePanel and any other GUI element that cares about that could update, rather than explicitly calling GearChangeFrema.updateFamiliars().

File: src/net/sourceforge/kolmafia/swingui/panel/GearChangePanel.java [hunk 2]
```diff
@@ -0,0 +1,1496 @@
...
+    this.outfitSelect =
+        new OutfitComboBox((LockableListModel<SpecialOutfit>) EquipmentManager.getOutfits());
+    this.customSelect =
+        new OutfitComboBox((LockableListModel<SpecialOutfit>) EquipmentManager.getCustomOutfits());
+
+    this.equipmentPanel = new EquipmentPanel();
+    this.customizablePanel = new CustomizablePanel();
+
+    this.tabs = KoLmafiaGUI.getTabbedPane();
+    this.tabs.setTabPlacement(SwingConstants.BOTTOM);
```
Comment (Veracity0) [inline src/net/sourceforge/kolmafia/swingui/panel/GearChangePanel.java:112]: Please use SwingConstants.TOP, as it was before.

**Review by Veracity0 — CHANGES_REQUESTED** [review]

File: src/net/sourceforge/kolmafia/swingui/panel/GearChangePanel.java [hunk 3]
```diff
@@ -0,0 +1,1496 @@
...
+      }
+    }
+  }
+
+  // For performance reasons, allow deferral of updateAllModels, which recomputes the lists that
+  // show in
+  // GearChangeFrame. This way we can perform multiple equipment updates at once without triggering
+  // several runs
+  // through the entire list of equipment in a player's inventory to see if it is equippable.
+  public static void deferUpdate() {
```
Comment (Veracity0) [inline src/net/sourceforge/kolmafia/swingui/panel/GearChangePanel.java:1269]: Could you fix the formatting of that comment, please?
Looks like spotlessapply broke long lines into two.

**Review by Veracity0 — CHANGES_REQUESTED** [review]

File: src/net/sourceforge/kolmafia/swingui/panel/GearChangePanel.java (hunk 2 above)
Comment (phulin) [reply src/net/sourceforge/kolmafia/swingui/panel/GearChangePanel.java:112]: Huh. I compared to my jar version of mafia to ensure it looked the same. The jar that I just downloaded has the tabs on the bottom. Maybe it's OS-dependent somehow? I don't see a preference that would make the positioning user-specific.

File: src/net/sourceforge/kolmafia/swingui/panel/GearChangePanel.java (hunk 1 above)
Comment (phulin) [reply src/net/sourceforge/kolmafia/swingui/panel/GearChangePanel.java:3]: thanks, this was my naughty IDE.

File: src/net/sourceforge/kolmafia/swingui/panel/GearChangePanel.java (hunk 3 above)
Comment (phulin) [reply src/net/sourceforge/kolmafia/swingui/panel/GearChangePanel.java:1269]: Will do.

File: src/net/sourceforge/kolmafia/swingui/panel/GearChangePanel.java (hunk 2 above)
Comment (Veracity0) [reply src/net/sourceforge/kolmafia/swingui/panel/GearChangePanel.java:137]: Java doc for JtabbedPane.setTabPlacement()

```public void setTabPlacement(int tabPlacement)```
Sets the tab placement for this tabbedpane. Possible values are:
- JTabbedPane.TOP
- JTabbedPane.BOTTOM
- JTabbedPane.LEFT
- JTabbedPane.RIGHT

The default value, if not set, is SwingConstants.TOP.

I have no idea why it is BOTTOM for you. Perhaps it is affected by look and feel?

GenericFrame does not call that - and GearChangeFrame has tabs at the top.
In your PR, I commented out the line - and the tabs are at the top.

I think you should remove that call and just let os/look-and-feel choose it.

Comment (phulin) [reply src/net/sourceforge/kolmafia/swingui/panel/GearChangePanel.java]: Done.

**Comment by phulin** [comment]
I think I've addressed all comments.

**Review by Veracity0 — APPROVED** [review]

**Pull request merged**